### ISSUE TITLE
"openEditor()" crashes when trying to focus existing webview 

### ISSUE THREAD

**Issue by Foxhoundn** [redhat-developer/vscode-extension-tester#316]
In one of our tests, the webview gets blurred and then the following is called

```
await webview.switchBack();
await editorView.openEditor('Qorus Webview');
await webview.switchToFrame();
```

But this fails with `StaleElementReferenceError: stale element reference: element is not attached to the page document`.

From the stack trace I see - ` From: Task: WebElement.findElement(By(xpath, .//div[starts-with(@id, 'webview-editor')]))`, but when I checked the VSCode HTML I don't see any element with `id` starting with `webview-editor`, only `data-parent-flow-to-element-id` exists with `webview-editor`.

Is this the correct API to use to switch to the open webview? (It used to work fine some months ago, we recently updated to 4.1.0).

Thanks

**Comment by jrichter1** [comment]
If I understand this correctly, you are trying to:
 1. open a web view and assign it to the `webview` object
 2. do whatever logic in the web view
 3. open a different web view and try to access it using the same `webview` object as before

The issue here might be the fact that you are reusing one object for two different web views (this probably was introduced in vscode 1.56 with the webview rework).
Try creating a new `WebView` instance after you open the second web view and use that one.

**Comment by Foxhoundn** [comment]
@jrichter1 there is just one webview, the `openEditor` call should just switch to the existing webview with title `Qorus Webview`, at least that's how it used to work.

**Comment by jrichter1** [comment]
The code starting with `await webview.switchBack();` implies that there is a web view before `'Qorus Webview'`is opened. And if it is the same web view, then the whole exercise of switching out only to switch back in seems useless.

Getting a `StaleElementReferenceError` means you are referencing an element whose identifier was changed, but not necessarily its locator or contents. Vscode does it quite a lot, reloading elements by deleting them from the DOM and adding them back in which generates different identifier for them.

The only reference I can see being used here is `webview`. It stands to reason the web view was reloaded between calls, particularly when the code in between is opening one.

**Comment by Foxhoundn** [comment]
Here is bit more of the code

```
export const createClass = async (webview: WebView, editorView: EditorView) => {
    await webview.switchBack();
    await editorView.openEditor('Qorus Webview');
    await webview.switchToFrame();
    await sleep(500);

    await clickElement(webview, 'button-create-new', 1, 'id');
    await sleep(500);

    await selectNthFolder(webview, 'target_dir', 1);
    await fillTextField(webview, 'field-class-class-name', 'ClassWithConfigItems');
    await fillTextField(webview, 'field-desc', 'Test class with config items');
    await fillTextField(webview, 'field-version', '1.3');

    await openConfigItemManager(webview);
    await openAddConfigItem(webview);

    // edit config item page
    await sleep(500);
    await fillTextField(webview, 'field-name', 'test-config-item-2');
    await fillTextField(webview, 'field-description', 'test config item of a base class');
    await selectField(webview, 'default_value');
```

The reason for the first `switchBack()` is because we do not know if we are currently in the webview or not (or rather not checking it), so we switch back and try to focus the webview again, even if it means we still were on the webview.

**Comment by jrichter1** [comment]
Alright, can I have the complete stacktrace at least?

**Comment by Foxhoundn** [comment]
Of course, here it is:

```
1) Config Items Tests
578       Create class:
579     StaleElementReferenceError: stale element reference: element is not attached to the page document
580  (Session info: chrome=89.0.4389.128)
581  (Driver info: chromedriver=89.0.4389.23 (61b08ee2c50024bab004e48d2b1b083cdbdac579-refs/branch-heads/4389@{#294}),platform=Linux 5.4.0-77-generic x86_64)
582      at Object.checkLegacyResponse (ui-test/node_modules/selenium-webdriver/lib/error.js:546:15)
583      at parseHttpResponse (ui-test/node_modules/selenium-webdriver/lib/http.js:509:13)
584      at doSend.then.response (ui-test/node_modules/selenium-webdriver/lib/http.js:441:30)
585      at process._tickCallback (internal/process/next_tick.js:68:7)
586  From: Task: WebElement.findElement(By(xpath, .//div[starts-with(@id, 'webview-editor')]))
587      at Driver.schedule (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
588      at WebView.schedule_ (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:2010:25)
589      at WebView.findElement (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:2057:17)
590      at WebView.<anonymous> (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:57:46)
591      at Generator.next (<anonymous>)
592      at /home/test/testing/omeoPQqUmXw2zQVLSTm1Akby/qorus-vscode/ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:8:71
593      at new Promise (<anonymous>)
594      at __awaiter (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:4:12)
595      at WebView.switchToFrame (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:52:16)
596      at /home/test/testing/omeoPQqUmXw2zQVLSTm1Akby/qorus-vscode/ui-test/out/tests/config_items.js:59:19
597  2) Config Items Tests
598       Create service:
599     StaleElementReferenceError: stale element reference: element is not attached to the page document
600  (Session info: chrome=89.0.4389.128)
601  (Driver info: chromedriver=89.0.4389.23 (61b08ee2c50024bab004e48d2b1b083cdbdac579-refs/branch-heads/4389@{#294}),platform=Linux 5.4.0-77-generic x86_64)
602      at Object.checkLegacyResponse (ui-test/node_modules/selenium-webdriver/lib/error.js:546:15)
603      at parseHttpResponse (ui-test/node_modules/selenium-webdriver/lib/http.js:509:13)
604      at doSend.then.response (ui-test/node_modules/selenium-webdriver/lib/http.js:441:30)
605      at process._tickCallback (internal/process/next_tick.js:68:7)
606  From: Task: WebElement.findElement(By(xpath, .//div[starts-with(@id, 'webview-editor')]))
607      at Driver.schedule (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:807:17)
608      at WebView.schedule_ (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:2010:25)
609      at WebView.findElement (ui-test/node_modules/selenium-webdriver/lib/webdriver.js:2057:17)
610      at WebView.<anonymous> (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:57:46)
611      at Generator.next (<anonymous>)
612      at /home/test/testing/omeoPQqUmXw2zQVLSTm1Akby/qorus-vscode/ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:8:71
613      at new Promise (<anonymous>)
614      at __awaiter (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:4:12)
615      at WebView.switchToFrame (ui-test/node_modules/monaco-page-objects/out/components/editor/WebView.js:52:16)
616      at /home/test/testing/omeoPQqUmXw2zQVLSTm1Akby/qorus-vscode/ui-test/out/tests/config_items.js:89:19
```

**Comment by noorshiblee** [comment]
I am also facing similar error ` Uncaught StaleElementReferenceError: stale element reference: element is not attached to the page document` .

    `await new Workbench().executeCommand('Open Home');
     await new Promise(res => {
         setTimeout(res,500);
     });
     view = new WebView();
     await view.switchToFrame();
     driver = VSBrowser.instance.driver;`

**Comment by jrichter1** [comment]
I've released an updated version of `WebView` in `4.1.1`. Tested it with a couple of different web view implementations, it seems to be working fine. If you can give it a try and tell me if it fixed your particular case, that would be great.

**Comment by orlade-anz** [comment]
Same issue for me, VS Code 1.59.0 and vscode-extension-tester 4.12 on Mac. There is an element with ID `webview-editor-element-<uuid>` but it wasn't where I expected to find it in the DOM. `$("div[id^=webview]")` in the console reveals it.

I'm using the `WebView` object, and `await webview.isDisplayed()` is `true`, but `await webview.switchToFrame();` fails.
